design.md

```diff
@@ -0,0 +1,157 @@
+---
+authors: Rex Posadas (rexposadas@gmail.com)
+---
+
+# What we are building
+
+Take-home challenge for Level 4 Systems Engineer.
+
+This documentation is a design for a job worker service and its associated CLI.
+
+The implementation is meant to be minimal while meeting all challenge requirements. 
+Tradeoffs are noted in the `tradeoffs` section of this document.
+
+# Parts of the challenge:
+
+## Worker Library
+
+1. Has methods to start/stop/query status and stream logs from the jobs.
+1. Resource controls of CPU, memory and disk IO using cgroups.
+
+## API
+
+1. gRPC API to start/stop/get status/stream output of a running process.
+1. Use mTLS for authentication.
+1. Has a simple authentication scheme.
+
+## Client
+
+1. CLI connects to worker service to start, stop, get status, and stream output of a job.
+2. Multiple clients can connect to the service.
+
+
+# Design
+
+## Worker Library and Server
+
+The server will use a library to: 
+
+1. Start/stop and query jobs. Starting a job returns a UUID which is the job id. This ID can be used to do further 
+   actions on the job. 
+2. Provide streaming of job logs.
+
+Service should be able to handle multiple clients. 
+
+*cgroups* will be used to limit resources. 
+- Each job will run in its own cgroup.
+- Resource usage will be isolated per job.
+- There will be job_id grouping. for example: 
+  - /sys/fs/cgroup/teleport/<job_id> -- the job_id is the UUID provided by the server when the job was created.
+
+
+## API
+
+### Using gRPC for communication. 
+
+- Start: Starts a job with hardcoded resource limits.
+- Stop: Stops a job.
+- GetStatus: Fetches status of a job
+- GetLogs: Stream logs of a running job.
+
+```
+service TeleportService {
+    rpc Start (StartRequest) returns (StartResponse);
+    rpc Stop (StopRequest) returns (StopResponse);
+    rpc GetStatus (GetStatusRequest) returns (GetStatusResponse);
+    rpc GetLogs (GetLogsRequest) returns (stream GetLogsResponse);
+}
+```
+
+### Authentication
+
+mTLS will be used for authentication. Both server and client provides certificates in order to 
+communicate. For testing, premade certificates will be provided.
+
+Users are authorized to use the service by its certification being validated.
+
+The CLI will look for certificates in the `~/.certs`.  These are the necessary files:
+
+Server:
+- `~/certs/ca.pem`
+- `~/certs/service.pem`
+- `~/certs/service-key.pem`
+
+Client:
+- `~/certs/ca.pem` 
+- `~/certs/client-<id>.pem` 
+- `~/certs/client-key-<id>.pem`
+
+
+#### Tradeoffs
+- The root CA are self-signed. Production applications should use a well-trusted CA.
+- The CLI only looks at `~/certs` for certificates. That's hardcoded.  Adding that as a CLI option would be easy enough, but didn't think it was necessary for this challenge. 
+
+
+### Streaming logs
+
+We will use a broadcasting model to stream logs to multiple clients.  Logs, stored in-memory, are
```

Comment:
A little more detail on how this will work would be good.

Reply:
Updated here: https://github.com/rexposadas/teleport/pull/1/commits/b0ecadcc492760cbb3be34e064c50f556eac83d3